v3/conn.go

```diff
@@ -145,35 +148,87 @@ type DialContext struct {
 	tc *tls.Config
 }
 
-func (dc *DialContext) dial(u *url.URL) (net.Conn, error) {
+func (dc *DialContext) dial(u *url.URL) (conn net.Conn, err error) {
 	if u.Scheme == "ldapi" {
 		if u.Path == "" || u.Path == "/" {
 			u.Path = "/var/run/slapd/ldapi"
 		}
 		return dc.d.Dial("unix", u.Path)
 	}
 
-	host, port, err := net.SplitHostPort(u.Host)
-	if err != nil {
-		// we assume that error is due to missing port
-		host = u.Host
-		port = ""
+	if u.Scheme != "ldap" && u.Scheme != "ldaps" {
+		return nil, fmt.Errorf("Unknown scheme '%s'", u.Scheme)
 	}
 
-	switch u.Scheme {
-	case "ldap":
-		if port == "" {
-			port = DefaultLdapPort
+	hostports := make([]string, 0, 1)
+
+	if u.Host == "" {
+		fragments := strings.Split(u.Path[1:], ",")
+		pieces := make([]string, 0, len(fragments))
+		for _, fragment := range fragments {
+			if strings.HasPrefix(fragment, "dc=") {
```

Comment:
A DS may use a different schema than Active Directory does. Novell eDirectory for example uses `o` instead of `dc`, which would break the code since you're only looking for a `dc=` prefix (even thought it's technically correct and would be [in line with the RFC 4519](https://datatracker.ietf.org/doc/html/rfc4519#section-2.4) to only consider `dc`). I think we're safe to assume that the URI won't be a full distinguished name to an organizational unit, like `ou=users,ou=site1,dc=my,dc=domain`. Therefore, I suggest we may change it to the following:

- Split the fragment on `=`
- Use index one of the result to get the name

This way we can safely ignore/not care about what schema the directory server is using.

Reply:
> Both JNDI LDAP and OpenLDAP client (e.g. `ldapsearch`) would lookup `_ldap._tcp.my.domain` for `ou=users,ou=site1,dc=my,dc=domain`, I don't want to break the kind of consent

Well, this means that if we want to stick to RFC 4519, then this implementation is ok to consider only the prefix "dc" 👍